### PR TITLE
Extend VoronoiFVM to incorporate FEM-computed velocity fields provided by ExtendableFEM

### PR DESCRIPTION
**Pull request by Da-Be-Ru**
We (@pjaap, @chmerdon and me) would like for `VoronoiFVM.jl` to admit FEM solutions as velocity fields in solving convection-diffusion systems.
For that purpose, we propose an extension `VoronoiFVMExtendableFEMBaseExt` which extends the capabilities of `edgevelocities` and `bfacevelocities`. They now provide the integrals of the velocity fields along Voronoi cell edges as computed by `ExtendableFEMBase`.

We also incorporated an example script which tests the fidelity of this extension and demonstrates the use case of working with two different grids: a flowgrid on which the velocity has been computed by `ExtendableFEM.jl` and a chemgrid onto which this velocity is interpolated to provide the input for the convection-diffusion system solved by `VoronoiFVM.jl`.

Mind that we still need the last [pull request](https://github.com/chmerdon/ExtendableFEMBase.jl/pull/3) for `ExtendableFEMBase.jl` to go through so we can adjust the `[compats]` for `VoronoiFVM.jl` accordingly.

### CODE REVIEW

**Comment by j-fu** [comment]
Thanks for the PR!
We need ExtendableFEM(Base) in test/Project.toml and docs/Project.toml .

**Comment by j-fu** [comment]
The current test in Example240 basically tests the if the interpolation from a FE space gives the same result as the interpolation of the analytical function. However this is no useful example to clone for users.

Therefore I propose the following:
- Move the following to a `test_feminterpolation.jl` in the tests subdirectory. It seems that this can run entirely without the solve. 
```
  @test norm(evelo_analytical .- evelo_fem, Inf) ≤ 1.0e-11
  @test norm(bfvelo_analytical .- bfvelo_fem, Inf) ≤ 1.0e-09
``` 
- Make the `full_fem_demo` the main in Example240. Possibly also compare to the analytical velocities. One could also test for the correct min/max.

**Comment by j-fu** [comment]
What are the `kwargs...` parameters doing ? They seem to be unused.

**Comment by j-fu** [comment]
Also please add docstrings to all newly introduced methods. 
In particular, the main API methods are `edgevelocities(grid, vel::FEVectorBlock)` and
`bfacevelocities(grid, vel::FEVectorBlock)`? Their docstrings will be for users, whereas the others would go to into the internal docs.

I would care about putting them into the Documenter build.

**Comment by pjaap** [comment]
> What are the `kwargs...` parameters doing ? They seem to be unused.

The new internal function `_integrate_along_segments` is not parameter free. It takes `interpolate_eps` and `axisymmetric` as kwargs. 
We want to be able to set this from the top level and therefore we need to hand down these kwargs through several functions.

Are we missing any possible side effects of this change?

**Comment by pjaap** [comment]
@j-fu All new methods are now documented. The original "internal" `integrate` calls for the `edgevelocities` have now a proper description in a docstring, too. 
However, they are documented in `Postprocessing` in the `VoronoiFVM` docs. This is probably not the optimal position.
Any suggestion where to put them? 
The different methods are already dispatched, so reorganizing the docs is simple.

**Comment by Da-Be-Ru** [comment]
Things are still not quite done. Here's a short list of issues that need work or resolving:
- getting the pull request for `ExtendableFEMBase.jl` merged
- somehow, we still get a non-vanishing FVM divergence for the cylindrical case in the Example script despite the fact that the computed velocity fields seems to match with the analytical solution at least on the grid nodes
- we discussed actually brute-forcing the `gFindLocal` call in `_integrate_along_segments` in case it comes up empty after the first call - we should still test that
- there's still the question of where to put the docs for the internal `integrate` functions; they should be *somewhere* as this is a huge help for anyone who needs to understand the internals
- since we now have a version of the example where the FEM-computed velocity field resolves the analytical solution exactly, we might not even need the old `main` anymore; @j-fu should we still put this into a separate test somewhere? This would be a test of the `edgevelocities` and `bfacevelocities` functionality without relying on the FEM solve at all, but just on the interpolation which is slimmer and more stable, but less useful for actual users as you point out

**Comment by j-fu** [comment]
> Things are still not quite done. Here's a short list of issues that need work or resolving:
> 
> * getting the pull request for `ExtendableFEMBase.jl` merged

Ok, this is done now.
> * somehow, we still get a non-vanishing FVM divergence for the cylindrical case in the Example script despite the fact that the computed velocity fields seems to match with the analytical solution at least on the grid nodes

How about the maximum principle ?  An the we also should cross-check the VoronoiFVM divergence caclulation
> * we discussed actually brute-forcing the `gFindLocal` call in `_integrate_along_segments` in case it comes up empty after the first call - we should still test that
ok
> * there's still the question of where to put the docs for the internal `integrate` functions; they should be _somewhere_ as this is a huge help for anyone who needs to understand the internals
This should go to the internal API section, see https://j-fu.github.io/VoronoiFVM.jl/stable/internal/
> * since we now have a version of the example where the FEM-computed velocity field resolves the analytical solution exactly, we might not even need the old `main` anymore; @j-fu should we still put this into a separate test somewhere? This would be a test of the `edgevelocities` and `bfacevelocities` functionality without relying on the FEM solve at all, but just on the interpolation which is slimmer and more stable, but less useful for actual users as you point out

I already created this test in my local copy. It will be available once I merge with your work, so you can delete this.

> 
> Oh, and note to @chmerdon: if we had a `HDIVBDM3` space, we could probably also resolve the quadratic case here 👀

Nice!

**Comment by chmerdon** [comment]
There is no HDIVBDM3, yet, but there is HDIVRTk(2,3)... will also have a look at the example next week.

**Comment by chmerdon** [comment]
Concerning the testcase that does not work with usefem = true and Cylindrical2D... I think the velocity from the fem computation is the reconstruction of r*v in that case, while the velocity for usefem = false or Cartesian2D is just v, right? Currently, it seems the FEM computes the divergence of (rv), which should be zero and is indeed zero, so only the divergence from the FVM is wrong. My guess is that the edgevelocities are probably computed incorrectly due to the wrong handling of the r*v input.

**Comment by chmerdon** [comment]
My guess is that this axisymmetric switch is not passed properly through from the edgevelocities function to the segment integration. Moreover, I think we should also provide the option to give the v instead of r*v also in the FEM case.

**Comment by Da-Be-Ru** [comment]
> My guess is that this axisymmetric switch is not passed properly through from the edgevelocities function to the segment integration. Moreover, I think we should also provide the option to give the v instead of r*v also in the FEM case.

You're correct, we forgot to handle to pass the `reconst = true` flag to the `edgevelocities` and `bfacevelocities` call for the cylindrical case. If it's not passed, we're using the wrong postprocessing kernel for the `SegmentIntegrator`.
With that fixed, the examples seem to run through as expected.

I'll push a cleaned up version with some more changes applied later.

**Comment by pjaap** [comment]
> An the we also should cross-check the VoronoiFVM divergence caclulation

I studied this function in order to understand the divergence problem and can say that it looks fine :)

**Comment by pjaap** [comment]
> This should go to the internal API section, see https://j-fu.github.io/VoronoiFVM.jl/stable/internal/

Done.

File: examples/Example240_FiniteElementVelocities.jl [hunk 1]
```diff
@@ -173,26 +137,35 @@ function full_fem_demo(;
...
     end
 
-    chemgrid[CoordinateSystem] = coord_system
+    if coord_system == Cylindrical2D
+        analytical_velocity = stagnation_flow_cylindrical
+    else
+        analytical_velocity = stagnation_flow_cartesian
+    end
+
+    if cylindrical_grid
```
Comment (pjaap) [inline examples/Example240_FiniteElementVelocities.jl:159]: Can you add a comment why you set this only for `chemgrid`?

File: examples/Example240_FiniteElementVelocities.jl [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+    data.cin = cin
+
+    evelo = edgevelocities(chemgrid, velocity; reconst)
+    bfvelo = bfacevelocities(chemgrid, velocity; reconst)
+
+    data.evelo = evelo
+    data.bfvelo = bfvelo
+
+    physics = VoronoiFVM.Physics(; flux = flux!, breaction = bconditions!, data)
+    sys = VoronoiFVM.System(chemgrid, physics; assembly = assembly)
```
Comment (pjaap) [inline examples/Example240_FiniteElementVelocities.jl:174]:
```suggestion
    sys = VoronoiFVM.System(chemgrid, physics; assembly)
```

File: examples/Example240_FiniteElementVelocities.jl [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+        @info "||div(R(v))||_2 = $(div_v)"
+    else
+        if cylindrical_grid
+            velocity = stagnation_flow_cylindrical
+        else
+            velocity = stagnation_flow_cartesian
+        end
+    end
+
+    if coord_system == Cylindrical2D
```
Comment (pjaap) [inline examples/Example240_FiniteElementVelocities.jl:153]:
```suggestion
    if cylindrical_grid
```

File: ext/VoronoiFVMExtendableFEMBaseExt.jl [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+
+            # proceed to next cell along edge of smallest barycentric coord
+            prevcell = icell
+            icell = xFaceCells[1, xCellFaces[facetogo[1][imin], icell]]
+            icell = icell == prevcell ? xFaceCells[2, xCellFaces[facetogo[1][imin], icell]] : icell
+
+            p1 .= pint
+        end
+
+        i += 1
```
Comment (pjaap) [inline ext/VoronoiFVMExtendableFEMBaseExt.jl:277]: What is `i`? Does it appear in the `while` loop? If so, give it a proper name :)

**Review by pjaap — COMMENTED** [review]
Thanks for fixes and cleanup. I left some nitpicking comments in the code :)

File: examples/Example240_FiniteElementVelocities.jl (hunk 3 above)
Comment (Da-Be-Ru) [reply examples/Example240_FiniteElementVelocities.jl:153]: Can't argue with that.

**Review by Da-Be-Ru — COMMENTED** [review]
Added comment as suggested.

File: examples/Example240_FiniteElementVelocities.jl (hunk 1 above)
Comment (pjaap) [reply examples/Example240_FiniteElementVelocities.jl:165]: It's an example. It is fine to have some verbose descriptions.

File: ext/VoronoiFVMExtendableFEMBaseExt.jl (hunk 4 above)
Comment (Da-Be-Ru) [reply ext/VoronoiFVMExtendableFEMBaseExt.jl:277]: `i` basically counts the number of steps through individual cells (=triangles) we've gone through in summing the integral through $\left\\{\lambda \\, \mathtt{p1} + (1-\lambda)\mathtt{p2} \\; : \\; \lambda\in[0,1]\right\\}$.

Used to be useful for debugging reasons, but you're right, it doesn't actually serve a purpose anymore.
I'd suggest getting rid of it.

**Pull request merged**